lib/util/git/auth.ts

```diff
@@ -0,0 +1,40 @@
+import { logger } from '../../logger';
+import { getHttpUrl } from './url';
+
+/*
+    Add authorization to a Git Url and returns the updated environment variables
+*/
+export function getGitAuthenticatedEnvironmentVariables(
+  gitUrl: string,
+  token: string,
+  environmentVariables?: NodeJS.ProcessEnv
+): NodeJS.ProcessEnv {
+  // check if the environmentVariables already contain a GIT_CONFIG_COUNT or if the process has one
+  const gitConfigCountEnvVariable =
+    environmentVariables?.GIT_CONFIG_COUNT || process.env.GIT_CONFIG_COUNT;
+  let gitConfigCount = 0;
+  if (gitConfigCountEnvVariable) {
+    // passthrough the gitConfigCountEnvVariable environment variable as start value of the index count
+    gitConfigCount = parseInt(gitConfigCountEnvVariable, 10);
+    if (Number.isNaN(gitConfigCount)) {
+      logger.warn(
+        `Found GIT_CONFIG_COUNT env variable, but couldn't parse the value to an integer: ${process.env.GIT_CONFIG_COUNT}. Ignoring it.`
+      );
+      gitConfigCount = 0;
+    }
+  }
+
+  const gitUrlWithToken = getHttpUrl(gitUrl, encodeURIComponent(token));
```

Comment:
Are you sure to encode the token here?
```suggestion
  const gitUrlWithToken = getHttpUrl(gitUrl, token);
```

I think `getHttpUrl` is already doing this for us.

Reply:
Not as far as I can see it, `getRemoteUrlWithToken()` as  well uses the same syntax: 
https://github.com/renovatebot/renovate/blob/3ffb57b1e8fa17360f1caf6c1c5fae97246f2a7a/lib/util/git/url.ts#L22

But I think we can remove the quote around the token in the `artifact.ts`:
https://github.com/renovatebot/renovate/blob/3ffb57b1e8fa17360f1caf6c1c5fae97246f2a7a/lib/manager/gomod/artifacts.ts#L30

Reply:
Can we verify that L13 did not already encode the token:
https://github.com/renovatebot/renovate/blob/38f8b55c738fb38c926bef8b716fe97eba105254/lib/util/git/url.ts#L5-L14

I also think github tokens don't need to be encoded at all.

Reply:
Line 13 does not correctly encode it. 
When trying it without the encoding it ended up looking like this, without the encoded `:`.
```
    -   "GIT_CONFIG_KEY_0": "url.https://token%3A1234@github.com/.insteadOf",
    +   "GIT_CONFIG_KEY_0": "url.https://token:1234@github.com/.insteadOf",
```
I added as well a test to ensure that it is correctly encoded.